Content/posts/otel-integration.md

```diff
@@ -0,0 +1,303 @@
+---
+date: 2024-12-16 14:00
+description: Getting started with Vapor 4 and OpenTelemetry
+tags: vapor, otel, prometheus, grafana
+image: /static/images/posts/otel-integration-arch.svg
+author: Paul
+authorImageURL: /author-images/paul.jpg
+---
+
+# Instrumenting Vapor 4 with Swift OTel
+
+This tutorial will guide you through setting up a Vapor 4 application with OpenTelemetry, collecting metrics about the application and its queue workers, and visualizing the data in Grafana.
+
+## Observability
+
+In complex systems it's often a good thing to be able to figure out why the system is behaving in a certain way without having to look into the code. When we want to be able to have a more high level view of what happens in the application without opening up the black box, we're looking for **observability**. Observability is the concept of collecting information about a system's execution and internal state, based on the data it generates.
+
+In more practical terms, observability is made up of 
+
+- **Logs**: exact details of an event, e.g an HTTP request, its method, time etc.;
+- **Metrics**: instant measurement representing some system state, such as number of HTTP requests/second;
+- **Traces**: a series of breadcrumbs which, if tied together, show the flow of data (such as a request) across the application, for instance how it gets routed across various internal components in the application.
+
+**Instrumenting** a system means adding observability capabilities. The idea is the code emits the data which then must be collected and sent to a backend observability system. 
+
+Before starting, let's give a small overview of our end goal. In Vapor 4, logging is integrated in the framework, and this post's aim is to add metrics collection to our system, but collecting tracing data is not much different. 
+
+![Vapor OTel Architecture](/static/images/posts/otel-integration-arch.svg)
+
+&nbsp;
+
+As you can see here there's a bunch of things going on, but we'll get into each of them. 
+
+1. Our Vapor HTTP server (which can be either an API or a web app) directly sends data to Prometheus, the metrics database;
```

Comment:
We need to clarify this (and the diagram). There's no functionality to send data to Prometheus, it's done the other way around, _something_ (normally a collector/side car) scrapes the metrics endpoint. I guess in our example it's Prometheus directly

Reply:
The way this works with an OTel Collector is that the server (Swift OTel) pushes signals (logs, metrics, spans) to the collector. The collector then either sends those signals to other tools, or in case of Prometheus for Metrics creates an HTTP server itself which Prometheus uses to scrape the metrics that were pushed to the collector.

Reply:
@slashmo That's the case for the OTel collector, however we're not using that for the HTTP server, we're using the collector only to collect metrics from the queues (this was a real project in which the queue metrics collection was an addition to the already established Vapor - Prometheus direct interaction solution). Since the HTTP server metrics are shared directly with Prometheus without going through the collector I believe Tim is right in saying that it's Prometheus fetching the data

Reply:
Yep, either way, Prometheus is always fetching.

Reply:
That being said, I was a bit surprised to see this shown in an OTel integration post. Is the goal of the post to show different ways of emitting metrics, one of which being OTel? Otherwise, if the goal is to integrate OTel, I'd expect Metrics and Tracing in there, with perhaps a couple of OTLP compatible observability backends so that folks know that OTel is more than just the collector.

Reply:
The post was born because we implemented this for a client and it seemed interesting to share, however I agree the title might not do it justice, we could update it to something like "Collecting metrics in Vapor with OTel and Prometheus". I haven't found any tutorial or blog post remotely similar to this so I guess anything's fine 😄